### PR TITLE
Reflect an user's `primary_email_address` update on its `contact_info` list

### PR DESCRIPTION
**Pull request by nicholaspcr**
#### Summary

Didn't create a issue for it but its linked to the list of behaviors noted while going through some of the `contact_info` changes. 

References the discussion in https://github.com/TheThingsNetwork/lorawan-stack/issues/6515#issuecomment-1737328616

Changes in the interactions between the update of an user `primary_email_address` and its `contact_info` list.

There are a few areas of which the intended behavior was not happening:
1. Effectively we do not update the primary email contact present in the `contact_info` list when an user performs a request to change its `primary_email_address`.
2. When a non admin updates the `primary_email_address`, it does not set the `primary_email_address_validated_at` to `nil`. It also does not send a email re-requesting its validation.

#### Changes

- Add `primary_email_address_validated_at` field mask when the update request is made by a non admin
- Updates now changes both the `primary_email_address` and its value present in the `contact_info` list.
- Add more unit test cases covering the sections of code changed 

#### Testing

Unit tests and manual tests.

<details>
  <summary> Test steps </summary>

  Run the Stack with the following configuration:
  ```yaml
  tls:
    source: file
    root-ca: ./ca.pem
    certificate: ./cert.pem
    key: ./key.pem

  is:
    email:
      dir: ".env/emails" # Example of directory path
      provider: "dir"
      sender-address: ""
      sender-name:  "employee@thethingsindustries.com" # Example "employee@thethingsindustries.com"
      templates:
        directory: # Example "<repo_path>/tts/pkg/email/templates"
    user-registration:
      contact-info-validation:
        required: true
        token-ttl: 48h0m0s
  ```

  Non admin:
  1. Create a new non admin user `ttn-lw-cli usr create <usr_id> --primary-email-address <email> --password <password>`
  2. Validate the new user's email (should have a validation email on the path in `dir` on the config)
  3. Run `tools/bin/mage dev:dbsql`
    3.1. Run `select primary_email_address, primary_email_address_validated_at from users;`
    3.2. User with <usr_id> should be validated
    3.3. Run `select * from contact_infos;`
    3.4. Contact info with <email> value should be validated.
  4. Login into the Console
  5. Access the `<base_url>/oauth/profile-settings` and update the `Email Address`
    5.1. Should receive a warning of restricted rights.
  6. Check the email folder, there should be a new validation email. This happens since the email is now not validated.
  7. Run `tools/bin/mage dev:dbsql`
    7.1. Run `select primary_email_address, primary_email_address_validated_at from users;`
    7.2. User with <usr_id> should be **not be** validated (validated_at being empty)
    7.3. Run `select * from contact_infos;`
    7.4. Contact info with <email> value should **not be** validated. (validated_at being empty)

  Admin:
  1. Create a new non admin user `ttn-lw-cli usr create <usr_id> --primary-email-address <email> --password <password> --admin`
  2. Validate the new user's email (should have a validation email on the path in `dir` on the config)
  3. Run `tools/bin/mage dev:dbsql`
    3.1. Run `select primary_email_address, primary_email_address_validated_at from users;`
    3.2. User with <usr_id> should be validated
    3.3. Run `select * from contact_infos;`
    3.4. Contact info with <email> value should be validated.
  4. Login into the Console
  5. Access the `<base_url>/oauth/profile-settings` and update the `Email Address`
  6. Run `tools/bin/mage dev:dbsql`
    6.1. Run `select primary_email_address, primary_email_address_validated_at from users;`
    6.2. User with <usr_id> should be **be** validated
    6.3. Run `select * from contact_infos;`
    6.4. Contact info with <email> value should **be** validated.

</details>

##### Regressions

Ever since https://github.com/TheThingsNetwork/lorawan-stack/commit/65edcafe193d9e266177e87cff977378f2cf8001 when an user who updates its `primary_email_address` the operation is not reflected on the `contact_info` list. Therefore there is a disconnect between the primary email and the values present in the `contact_info`. The field has been deprecated, the primary email is what is used to send emails and we removed references to the `contact_info` field, being only kept for consistency.

Therefore I believe that this change is merely superficial and done in order to have the behavior as expected until the next `major/bump` and we can remove the `contact_info` field. 

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] The steps/process to test this feature are clearly explained including testing for regressions.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by KrishnaIyer** [comment]
@nicholaspcr: Please add the testing procedure.

File: pkg/identityserver/user_registry.go [hunk 1]
```diff
@@ -478,39 +481,60 @@ func (is *IdentityServer) updateUser(ctx context.Context, req *ttnpb.UpdateUserR
...
-				}
+		updatingContactInfo := ttnpb.HasAnyField(req.FieldMask.GetPaths(), "contact_info")
+
+		contactInfo, err := st.GetContactInfo(ctx, req.User.GetIds())
+		if err != nil {
+			return err
+		}
+
+		if updatingPrimaryEmailAddress {
+			oldUser, err := st.GetUser(ctx, req.User.GetIds(), []string{"primary_email_address"})
```
Comment (adriansmares) [inline pkg/identityserver/user_registry.go:502]: This assumes that the previous primary email address will be found in the `contact_info`, which is not the case for users which have change their primary email address in the meantime.

An update to the `primary_email_address` without an update to `contact_info` should forcefully re-create the `contact_info` using the new email address in my opinion.

Comment (nicholaspcr) [reply pkg/identityserver/user_registry.go:502]: It was more under the assumption that it shouldn't be done but took some time to do some changes which I believe it makes more sense.

**Comment by nicholaspcr** [comment]
>The contact_info field even if deprecated, it being removed only on the major bump implies in the following scenario being possible in the user's update operation: an user updates primary_email_address and contact_info at the same time, if the new primary_email_address is not on the new contact_info values then there won't be an contact for the new primary_email_address value.

With the new changes done this should not longer be the case so I'm removing the text from the note for reviewers.

File: pkg/identityserver/user_registry.go (hunk 1 above)
Comment (adriansmares) [reply pkg/identityserver/user_registry.go:502]: The problem here is that there no guarantee that the contact info email is the old user email basically. If I edited my email on version `v3.27.2`, from `adrian@thethingsindustries.com` to `adrian.mares@thethingsindustries.com`, my email in `oldUser.PrimaryEmailAddress` is `adrian.mares@thethingsindustries.com` while the contact info will contain `adrian@thethingsindustries.com` - the loop bellow won't find it and will just append a new email in the contact info.

You cannot depend on the contact info containing anything related to the 'old' user - you should just always override it as long as the user didn't update it himself.

File: pkg/identityserver/user_registry.go [hunk 2]
```diff
@@ -533,6 +571,14 @@ func (is *IdentityServer) updateUser(ctx context.Context, req *ttnpb.UpdateUserR
 		})
 	}
 
+	// If primary email address was updated and invalidated, send a new validation email.
+	if usr.PrimaryEmailAddressValidatedAt == nil &&
```
Comment (adriansmares) [inline pkg/identityserver/user_registry.go]: The trigger here should be the contact info change, not the `PrimaryEmailAddressValidatedAt`, as the contact info is what is validated (and ends up also updating `PrimaryEmailAddressValidatedAt`.

File: pkg/identityserver/user_registry.go (hunk 1 above)
Comment (nicholaspcr) [reply pkg/identityserver/user_registry.go:502]: Agree but that is only for the first update after this PR is merged, since https://github.com/TheThingsNetwork/lorawan-stack/blob/b956d4a6af012dab137d4ef0f5942bd6b510fb52/pkg/identityserver/user_registry.go#L537-L547 ensures that if the new email will be added to the contactInfo list if the old email was not, all subsequent operations will have a oldEmail to look into within the contactInfo list.

Comment (adriansmares) [reply pkg/identityserver/user_registry.go:502]: But this is my problem here - there is no point in keeping the old contact info in place. It's an old, unused email which shouldn't be stored - we should just override it.

File: pkg/identityserver/user_registry.go (hunk 2 above)
Comment (nicholaspcr) [reply pkg/identityserver/user_registry.go]: not sure if I agree with this, the way I see this is that an emailValidateAt will not be `nil` if:
1. An admin is updating it and sets the `PrimaryEmailAddressValidatedAt` in the request body.
2. An admin updates just the `PrimaryEmailAddress` field and expects for it to be validated since the old email was already validated.


But it can be `nil` if an admin did the update of `PrimaryEmailAddress` before this fix and therefore its value its not on the `contactInfo` list, after v3.28 a new update of just `PrimaryEmailAddress` field is done making the following true:
  1.  no old contact is found
  2. there is no `validateAt` for the new email to inherit
  3. There won't be a `ContactInfoChange` as true and the new email will be added to the `contactInfo` list
  4. The new email is invalidated and has to be validated.

I believe that having the `PrimaryEmailAddressValidatedAt == nil` covers this case, which is an edge case but it has a good change of happening.

File: pkg/identityserver/user_registry.go (hunk 1 above)
Comment (nicholaspcr) [reply pkg/identityserver/user_registry.go:502]: I see, I was trying to avoid overwriting the data in the field until we remove everything in https://github.com/TheThingsNetwork/lorawan-stack/issues/6515 but if its permissible to do so then I agree on just overwriting it

File: pkg/identityserver/user_registry.go [hunk 3]
```diff
@@ -432,7 +431,11 @@ func (is *IdentityServer) updateUser(ctx context.Context, req *ttnpb.UpdateUserR
 
 	if !updatedByAdmin {
 		req.User.PrimaryEmailAddressValidatedAt = nil
-		cleanContactInfo(req.User.ContactInfo)
+		req.FieldMask.Paths = ttnpb.AddFields(req.FieldMask.GetPaths(), "primary_email_address_validated_at")
+		if len(req.User.ContactInfo) > 0 {
+			cleanContactInfo(req.User.ContactInfo)
+			req.FieldMask.Paths = ttnpb.AddFields(req.FieldMask.GetPaths(), "contact_info")
```
Comment (adriansmares) [inline pkg/identityserver/user_registry.go]: I would prefer that we don't add new fields to the field mask in this manner. This makes the `contact_info` field mask implicit if the value is not empty and the user is not an admin, which is not how field masks are supposed to work.

File: pkg/identityserver/user_registry.go (hunk 2 above)
Comment (adriansmares) [reply pkg/identityserver/user_registry.go]: It's more of a consistency thing - since `contact_info` is a list, any item of that list could have `validated_at` set to `nil` and would require a contact info validation. I am ok with keeping this as proposed here and getting rid of the mutable contact info ASAP.

File: pkg/identityserver/user_registry.go (hunk 3 above)
Comment (nicholaspcr) [reply pkg/identityserver/user_registry.go]: While I agree with the sentiment and the conditional of `contact_info` not being an empty slice is unnecessary I do wonder if it is possible to enforce non admin behaviors without affecting the field mask minimally. 

For example, this feels acceptable to me
```go
req.FieldMask.Paths = ttnpb.AddFields(req.FieldMask.GetPaths(), "primary_email_address_validated_at")
```
but indeed the 
```go
if len(req.User.ContactInfo) > 0 {
	cleanContactInfo(req.User.ContactInfo)
	req.FieldMask.Paths = ttnpb.AddFields(req.FieldMask.GetPaths(), "contact_info")
```
is unecessary.

Comment (adriansmares) [reply pkg/identityserver/user_registry.go]: The validate at timestamp is indeed a fine addition because we provide the value intentionally - we control the value put into the request on the RPC side, so the intent is clear.

The `contact_info` here is added just because the slice is non empty in the request, even though that slice comes from the caller - that is what I find wrong here. Caller input must be provided next to a field mask entry, otherwise it must always be ignored.

File: pkg/identityserver/user_registry.go (hunk 2 above)
Comment (nicholaspcr) [reply pkg/identityserver/user_registry.go]: Left it with pretty much the same logic since the removal of the mutable contact info is the priority after the notification API, so I believe it is okay as is. There is a comment for more clarification around it

File: pkg/identityserver/user_registry.go [hunk 4]
```diff
@@ -471,48 +471,80 @@ func (is *IdentityServer) updateUser(ctx context.Context, req *ttnpb.UpdateUserR
...
+		if updatingContactInfo {
+			contactInfo, err = st.SetContactInfo(ctx, req.User.GetIds(), req.User.ContactInfo)
+			if err != nil {
+				return err
+			}
+		}
+
+		if updatingPrimaryEmailAddress {
+			// If no update was done to contactInfo list then fetch it.
+			if contactInfo == nil {
```
Comment (adriansmares) [inline pkg/identityserver/user_registry.go]: The real condition here is `!updatingContactInfo`, right ? It is possible that the user patched with a `nil` `contact_info`.

File: pkg/identityserver/user_registry.go [hunk 5]
```diff
@@ -471,48 +471,80 @@ func (is *IdentityServer) updateUser(ctx context.Context, req *ttnpb.UpdateUserR
...
+			if err != nil {
+				return err
+			}
 		}
+
 		usr, err = st.UpdateUser(ctx, req.User, req.FieldMask.GetPaths())
 		if err != nil {
 			return err
 		}
-		if updatingContactInfo {
```
Comment (adriansmares) [inline pkg/identityserver/user_registry.go:513]: Why was this removed ?

Comment (nicholaspcr) [reply pkg/identityserver/user_registry.go:513]: we update the contact info before updating the user and we are already fetching the contact info associated with the user. At least that is what I understood from the usage of the `bun.Relation` method, plus the tests do validate what is returned. So the removal of the conditional seems appropriate.

Comment (adriansmares) [reply pkg/identityserver/user_registry.go:513]: I see, you are right. 

I think we should remove the `contact_info` from the `UpdateUser` call field mask to avoid patching twice. Or we could just try to work with a single `UpdateUser` if possible.

Comment (nicholaspcr) [reply pkg/identityserver/user_registry.go:513]: Never-mind, indeed overwriting the `usr.ContactInfo` was necessary, the unit tests did a `Get` after the update and validated that, which is by itself a mistake.

Nevertheless the idea of removing the `contact_info` from the field mask before the `userUpdate` is still valid and I opted to add it in the simplest form. Will attempt to do all of the updates within a single `updateUser` call when working on the `contact_info` store removal.

**Review by adriansmares — APPROVED** [review]
